### ISSUE TITLE
TF: Dropout still applies on inference if training = False

### ISSUE THREAD

**Issue by amurashov** [lmnt-com/haste#18]
I think it affects all layers which have 'dropout' - on inference
``
tf.nn.dropout[weights['recurrent_kernel'], rate = self.dropout)
``
is passed onto your C++ / CUDA part. Smth of this sort is more appropriate:
``
tf.nn.dropout[weights['recurrent_kernel'], rate = (self.dropout if training else 0.0))
``

I am not sure if zoneout is applied correctly under the hood too, does 'training' parameter to CUDA part switches it off correctly? I hacked around it amending relevant code as following:
``
h, c, _ = LIB.haste_lstm(
        x,
        weights['kernel'],
        tf.nn.dropout(weights['recurrent_kernel'], rate = (self.dropout if training else 0.0)),
        weights['bias'],
        zoneout_mask if training else tf.zeros([0, 0, 0], dtype=self.dtype),
        training=training,
        zoneout_prob=(self.zoneout if training else 0.0))
``

**Comment by amurashov** [comment]
I ran some experiements and it seems zoneout applies too on inference, so looks like a bug in the C++ / CUDA part too.

**Comment by sharvil** [comment]
Great catch on dropout! I'll send out a fix once I've verified it.

Zoneout is applied correctly. The computation looks like:

```cpp
  if (ApplyZoneout) {
    if (Training) {
      cur_h_value = (cur_h_value - h[output_idx]) * zoneout_mask[output_idx] + h[output_idx];
    } else {
      cur_h_value = (zoneout_prob * h[output_idx]) + ((1.0f - zoneout_prob) * cur_h_value);
    }
  }
```

The zoneout mask is ignored, and there's a weighted average of the previous cell state and current cell state based on `zoneout_prob`. That's the correct way to apply zoneout during inference to match the distribution during training.

**Comment by amurashov** [comment]
Does it mean that if we just 'copy' (rather transfer as format is different) weights trained in HASTE with Zoneout != 0 to production (e.g. 'standard' LSTMs) to deploy it will be not correct?

Dropout layers scale activations during training to ensure that if on inference you just skip dropout layer - the inference is correct, isn't the same possible with Zoneout?

**Comment by sharvil** [comment]
Re: dropout. Yes, if you transfer the weights to another system you won't get the correct result with the current implementation unless you also scale the recurrent matrix by `1/(1-dropout)`.

Re: zoneout. Scaling zoneout activations won't produce correct results. Note that dropout sets activations to 0 whereas zoneout copies activations from the previous timestep. Dropout is data-independent whereas zoneout is data-dependent.

Consider a situation where state `h(t-1)` has distribution A and `h(t)` has distribution B. During training with `zoneout_prob > 0`, `h(t)` has seen samples from distribution A and distribution B. We need to make sure that at test time, `h(t)` still sees samples from distribution A.

FWIW, code written by the authors of the Zoneout paper also do it this way: https://github.com/teganmaharaj/zoneout/blob/master/zoneout_tensorflow.py#L54
